phpBB/phpbb/messenger/method/phpbb_email.php

```diff
@@ -0,0 +1,637 @@
+	}
+
+	/**
+	 * Set the reply to address
+	 *
+	 * @param string	$address	Email "Reply to" address
+	 * @return void
+	 */
+	public function replyto($address)
+	{
+		$this->replyto = new Address(trim($address));
+		$this->email->getReplyTo() ? $this->email->addReplyTo($this->replyto) : $this->email->replyTo($this->replyto);
+	}
+
+	/**
+	 * Set the from address
+	 *
+	 * @param string	$address	Email "from" address
+	 * @return void
+	 */
+	public function from($address)
+	{
+		$this->from = new Address(trim($address));
+		$this->email->getFrom() ? $this->email->addFrom($this->from) : $this->email->from($this->from);
+	}
+
+	/**
+	 * Set up subject for mail
+	 *
+	 * @param string	$subject	Email subject
+	 * @return void
+	 */
+	public function subject($subject = '')
+	{
+		parent::subject(trim($subject));
+		$this->email->subject($this->subject);
+	}
+
+	/**
+	 * Set up extra mail headers
+	 *
+	 * @param string	$header_name	Email header name
+	 * @param string	$header_value	Email header body
+	 * @return void
+	 */
+	public function header($header_name, $header_value)
+	{
+		$header_name = $header_name;
+		$header_value = $header_value;
+
+		// addMailboxListHeader() requires value to be array
+		if ($this->get_header_method($header_name) == 'addMailboxListHeader')
+		{
+			$header_value = [$header_value];
+		}
+		$this->headers->addHeader($header_name, $header_value);
+	}
+
+	/**
+	 * Adds X-AntiAbuse headers
+	 *
+	 * @param \phpbb\config\config	$config		Config object
+	 * @param \phpbb\user			$user		User object
+	 * @return void
+	 */
+	public function anti_abuse_headers($config, $user)
+	{
+		$this->header('X-AntiAbuse', 'Board servername - ' . $config['server_name']);
+		$this->header('X-AntiAbuse', 'User_id - ' . $user->data['user_id']);
+		$this->header('X-AntiAbuse', 'Username - ' . $user->data['username']);
+		$this->header('X-AntiAbuse', 'User IP - ' . $user->ip);
+	}
+
+	/**
+	 * Set the email priority
+	 *
+	 * Possible values are:
+	 * Email::PRIORITY_HIGHEST = 1
+	 * Email::PRIORITY_HIGH = 2
+	 * Email::PRIORITY_NORMAL = 3
+	 * Email::PRIORITY_LOW = 4
+	 * Email::PRIORITY_LOWEST = 5
+	 *
+	 * @param int	$priority	Email priority level
+	 * @return void
+	 */
+	public function set_mail_priority($priority = Email::PRIORITY_NORMAL)
+	{
+		$this->email->priority($priority);
+	}
+
+	/**
+	 * Detect proper Header class method to add header
+	 *
+	 * @param string	$name	Email header name
+	 * @return string
+	 */
+	protected function get_header_method(string $name)
+	{
+		$parts = explode('\\', self::HEADER_CLASS_MAP[strtolower($name)] ?? UnstructuredHeader::class);
+		$method = 'add'.ucfirst(array_pop($parts));
+		if ('addUnstructuredHeader' === $method)
+		{
+			$method = 'addTextHeader';
+		}
+		else if ('addIdentificationHeader' === $method)
+		{
+			$method = 'addIdHeader';
+		}
+
+		return $method;
+	}
+
+	/**
+	 * Set email headers
+	 *
+	 * @return bool
+	 */
+	protected function build_header()
```

Comment:
So the whole point of this is just to allow modification via the event?

Reply:
Yes but also to build a preset of headers.